src/compiler/checker.ts

```diff
@@ -6653,23 +6653,23 @@ namespace ts {
             }
 
             function getPropertyNameNodeForSymbol(symbol: Symbol, context: NodeBuilderContext) {
+                const stringNamed = !!length(symbol.declarations) && every(symbol.declarations, isStringNamed);
                 const singleQuote = !!length(symbol.declarations) && every(symbol.declarations, isSingleQuotedStringNamed);
-                const fromNameType = getPropertyNameNodeForSymbolFromNameType(symbol, context, singleQuote);
+                const fromNameType = getPropertyNameNodeForSymbolFromNameType(symbol, context, singleQuote, stringNamed);
                 if (fromNameType) {
                     return fromNameType;
                 }
                 const rawName = unescapeLeadingUnderscores(symbol.escapedName);
-                const stringNamed = !!length(symbol.declarations) && every(symbol.declarations, isStringNamed);
                 return createPropertyNameNodeForIdentifierOrLiteral(rawName, getEmitScriptTarget(compilerOptions), singleQuote, stringNamed);
             }
 
             // See getNameForSymbolFromNameType for a stringy equivalent
-            function getPropertyNameNodeForSymbolFromNameType(symbol: Symbol, context: NodeBuilderContext, singleQuote?: boolean) {
+            function getPropertyNameNodeForSymbolFromNameType(symbol: Symbol, context: NodeBuilderContext, singleQuote?: boolean, stringNamed?: boolean) {
                 const nameType = getSymbolLinks(symbol).nameType;
                 if (nameType) {
                     if (nameType.flags & TypeFlags.StringOrNumberLiteral) {
                         const name = "" + (nameType as StringLiteralType | NumberLiteralType).value;
-                        if (!isIdentifierText(name, getEmitScriptTarget(compilerOptions)) && !isNumericLiteralName(name)) {
+                        if (!isIdentifierText(name, getEmitScriptTarget(compilerOptions)) && (stringNamed || !isNumericLiteralName(name))) {

```

Comment:
Would it make sense to just rewrite this as

```suggestion
                        if (stringNamed || !isIdentifierText(name, getEmitScriptTarget(compilerOptions))) {
```

Reply:
Nah, we need the numeric name callout, otherwise we don't prefer writing `[-1]:` over `"-1":`, or `0:` over `"0":` (which is important, since our types for `keyof` key off weather those are strings or not, so we really wanna prefer the non-string form unless it's overridden).